<h1><b><p style="background-image: url(https://i.postimg.cc/0Qwf8YX6/2927262.jpg);font-family:camtasia;font-size:110%;color:white;text-align:center;border-radius:15px 50px; padding:7px; border:solid 2px #09375b; box-shadow: 10px 10px 10px #042b4c">Project Title: Social Media Usage and Emotional Well-Being</p></b></h1>

# About Dataset
This unique dataset was meticulously researched and prepared by AI Inventor Emirhan BULUT. It captures valuable information on social media usage and the dominant emotional state of users based on their activities. The dataset is ideal for exploring the relationship between social media usage patterns and emotional well-being.
Features:

Files:

    train.csv: Data for training models.
    test.csv: Data for testing models.
    val.csv: Data for validation purposes.

__Dataset Overview:__

  🌐 Source: __Dataset is taken from [Kaggle](https://www.kaggle.com/datasets/emirhanai/social-media-usage-and-emotional-well-being/data)__

* __Columns:__
  * User_ID: Unique identifier for the user.
  * Age: Age of the user.
  * Gender: Gender of the user (Female, Male, Non-binary).
  * Platform: Social media platform used (e.g., Instagram, Twitter, Facebook, LinkedIn, Snapchat, Whatsapp, Telegram).
  * Daily_Usage_Time (minutes): Daily time spent on the platform in minutes.
  * Posts_Per_Day: Number of posts made per day.
  * Likes_Received_Per_Day: Number of likes received per day.
  * Comments_Received_Per_Day: Number of comments received per day.
  * Messages_Sent_Per_Day: Number of messages sent per day.
  * Dominant_Emotion: User's dominant emotional state during the day (e.g., Happiness, Sadness, Anger, Anxiety, Boredom, Neutral).

__Attribution:__

You can use this dataset ethically and responsibly in accordance with the MIT license for educational and research purposes. If you use this dataset in your work, a citation or reference to this dataset would be appreciated.

# Introduction:
Social media is a central point of our life. It has changed the way we live, work, and interact with others. As a result, it has become an essential part of our daily lives. The dataset here aims to provide valuable insights into social media usage and the dominant emotional state of users.

The goal of this project is to analyze the relation between social media trends and emotional well-being, and train the nodels.

### About Author:
__Author: *Saadat Khalid Awan*__

| **Contact Information** | |
|-------------------------|---------------------------------------------------|
| **Email:** | *me.saadi96@gmail.com* |
| **Website:** | *[https://thesaadat.blogspot.com](https://thesaadat.blogspot.com)* |

__🌐 Let's Connect:__

| Platform   | Badge                                                                                                  | URL                                                      |
|------------|--------------------------------------------------------------------------------------------------------|----------------------------------------------------------|
| Facebook   | ![Facebook](https://img.shields.io/badge/Facebook-%231877F2.svg?logo=Facebook&logoColor=white)        | [Saadat.Khalid.Awan](https://facebook.com/Saadat.Khalid.Awan) |
| Instagram  | ![Instagram](https://img.shields.io/badge/Instagram-%23E4405F.svg?logo=Instagram&logoColor=white)     | [saadii_awan66](https://instagram.com/saadii_awan66)           |
| LinkedIn   | ![LinkedIn](https://img.shields.io/badge/LinkedIn-%230077B5.svg?logo=linkedin&logoColor=white)        | [saadatawan](https://linkedin.com/in/saadatawan)              |
| Medium     | ![Medium](https://img.shields.io/badge/Medium-12100E?logo=medium&logoColor=white)                     | [@me.saadat](https://medium.com/@@me.saadat)                  |
| Pinterest  | ![Pinterest](https://img.shields.io/badge/Pinterest-%23E60023.svg?logo=Pinterest&logoColor=white)     | [its_saadatkhalid](https://pinterest.com/its_saadatkhalid)    |
| Quora      | ![Quora](https://img.shields.io/badge/Quora-%23B92B27.svg?logo=Quora&logoColor=white)                 | [Saadat-Khalid-Awan](https://quora.com/profile/Saadat-Khalid-Awan) |
| TikTok     | ![TikTok](https://img.shields.io/badge/TikTok-%23000000.svg?logo=TikTok&logoColor=white)              | [saadat.awan](https://tiktok.com/@saadat.awan)                 |
| Twitter    | ![Twitter](https://img.shields.io/badge/Twitter-%231DA1F2.svg?logo=Twitter&logoColor=white)           | [saadat_96](https://twitter.com/saadat_96)                    |
| YouTube    | ![YouTube](https://img.shields.io/badge/YouTube-%23FF0000.svg?logo=YouTube&logoColor=white)           | [saadatkhalidawan](https://youtube.com/@saadatkhalidawan)     |
| Github     | ![Github](https://img.shields.io/badge/Github-%23FF0000.svg?logo=Github&logoColor=Black)              | [Saadat-Khalid](https://github.com/Saadat-Khalid/)           |

# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
# Ignore Warnings
import warnings
warnings.filterwarnings("ignore")

# Data Loading

* We've 3 files:
  * test.csv
  * train.csv
  * val.csv

In [3]:
train_df = pd.read_csv("/kaggle/input/social-media-usage-and-emotional-well-being/train.csv", on_bad_lines='skip')
test_df = pd.read_csv("/kaggle/input/social-media-usage-and-emotional-well-being/test.csv", on_bad_lines='skip')
val_df = pd.read_csv("/kaggle/input/social-media-usage-and-emotional-well-being/val.csv", on_bad_lines='skip')

* Before diving into the data, let's take a look at the data. Afterwards we'll explore the traning dataset.

In [4]:
# viwing the traing data, test data and validation data
print("Training Data:")
display(train_df.head())
print("----------------------------------------------------------------------")
print("Testing Data:")
display(test_df.head())
print("----------------------------------------------------------------------")
print("Validation Data:")
display(val_df.head())

Training Data:


User_ID Age      Gender   Platform  Daily_Usage_Time (minutes)  \
0       1  25      Female  Instagram                       120.0   
1       2  30        Male    Twitter                        90.0   
2       3  22  Non-binary   Facebook                        60.0   
3       4  28      Female  Instagram                       200.0   
4       5  33        Male   LinkedIn                        45.0   

   Posts_Per_Day  Likes_Received_Per_Day  Comments_Received_Per_Day  \
0            3.0                    45.0                       10.0   
1            5.0                    20.0                       25.0   
2            2.0                    15.0                        5.0   
3            8.0                   100.0                       30.0   
4            1.0                     5.0                        2.0   

   Messages_Sent_Per_Day Dominant_Emotion  
0                   12.0        Happiness  
1                   30.0            Anger  
2                   20.0          Neutral  
3                   50.0          Anxiety  
4                   10.0          Boredom

----------------------------------------------------------------------
Testing Data:


User_ID Age      Gender  Platform  Daily_Usage_Time (minutes)  \
0      500  27      Female  Snapchat                         120   
1      488  21  Non-binary  Snapchat                          60   
2      776  28  Non-binary  Snapchat                         115   
3      869  27        Male  Telegram                         105   
4      573  21  Non-binary  Facebook                          55   

   Posts_Per_Day  Likes_Received_Per_Day  Comments_Received_Per_Day  \
0              4                      40                         18   
1              1                      18                          7   
2              3                      38                         18   
3              3                      48                         20   
4              3                      17                          7   

   Messages_Sent_Per_Day Dominant_Emotion  
0                     22          Neutral  
1                     12          Neutral  
2                     27          Anxiety  
3                     28          Anxiety  
4                     12          Neutral

----------------------------------------------------------------------
Validation Data:


User_ID Age      Gender   Platform  Daily_Usage_Time (minutes)  \
0       10  31        Male  Instagram                         170   
1      877  32      Female  Instagram                         155   
2      230  26  Non-binary   Facebook                          45   
3      876  28  Non-binary   Snapchat                         115   
4      376  28  Non-binary   Snapchat                         115   

   Posts_Per_Day  Likes_Received_Per_Day  Comments_Received_Per_Day  \
0              5                      80                         20   
1              6                      75                         25   
2              1                       8                          4   
3              3                      38                         18   
4              3                      38                         18   

  Messages_Sent_Per_Day Dominant_Emotion  
0                    35        Happiness  
1                    38        Happiness  
2                    12          Sadness  
3                    27          Anxiety  
4                    27          Anxiety

In [5]:
# training data info
print("Traing Data Info:")
train_df.info()

Traing Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   User_ID                     1001 non-null   object 
 1   Age                         1001 non-null   object 
 2   Gender                      1000 non-null   object 
 3   Platform                    1000 non-null   object 
 4   Daily_Usage_Time (minutes)  1000 non-null   float64
 5   Posts_Per_Day               1000 non-null   float64
 6   Likes_Received_Per_Day      1000 non-null   float64
 7   Comments_Received_Per_Day   1000 non-null   float64
 8   Messages_Sent_Per_Day       1000 non-null   float64
 9   Dominant_Emotion            1000 non-null   object 
dtypes: float64(5), object(5)
memory usage: 78.3+ KB


In [6]:
# testing data info
print("Testing Data Info:")
test_df.info()

Testing Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   User_ID                     103 non-null    int64 
 1   Age                         103 non-null    object
 2   Gender                      103 non-null    object
 3   Platform                    103 non-null    object
 4   Daily_Usage_Time (minutes)  103 non-null    int64 
 5   Posts_Per_Day               103 non-null    int64 
 6   Likes_Received_Per_Day      103 non-null    int64 
 7   Comments_Received_Per_Day   103 non-null    int64 
 8   Messages_Sent_Per_Day       103 non-null    int64 
 9   Dominant_Emotion            103 non-null    object
dtypes: int64(6), object(4)
memory usage: 8.2+ KB


In [7]:
# validation data info
print("Validation Data Info:")
val_df.info()

Validation Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145 entries, 0 to 144
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   User_ID                     145 non-null    int64 
 1   Age                         145 non-null    object
 2   Gender                      145 non-null    object
 3   Platform                    145 non-null    object
 4   Daily_Usage_Time (minutes)  145 non-null    int64 
 5   Posts_Per_Day               145 non-null    int64 
 6   Likes_Received_Per_Day      145 non-null    int64 
 7   Comments_Received_Per_Day   145 non-null    int64 
 8   Messages_Sent_Per_Day       145 non-null    object
 9   Dominant_Emotion            144 non-null    object
dtypes: int64(5), object(5)
memory usage: 11.5+ KB


* Shape of the training, testing, and validation data.

In [8]:
print(f"There are {train_df.shape[0]} rows and {train_df.shape[1]} columns in the training data.")
print(f"There are {test_df.shape[0]} rows and {test_df.shape[1]} columns in the testing data.")
print(f"There are {val_df.shape[0]} rows and {val_df.shape[1]} columns in the validation data.")

There are 1001 rows and 10 columns in the training data.
There are 103 rows and 10 columns in the testing data.
There are 145 rows and 10 columns in the validation data.


In [9]:
# checking for null values
print("Training Data:")
display(train_df.isnull().sum())
print("----------------------------------------------------------------------")
print("Testing Data:")
display(test_df.isnull().sum())
print("----------------------------------------------------------------------")
print("Validation Data:")
display(val_df.isnull().sum())

Training Data:


User_ID                       0
Age                           0
Gender                        1
Platform                      1
Daily_Usage_Time (minutes)    1
Posts_Per_Day                 1
Likes_Received_Per_Day        1
Comments_Received_Per_Day     1
Messages_Sent_Per_Day         1
Dominant_Emotion              1
dtype: int64

----------------------------------------------------------------------
Testing Data:


User_ID                       0
Age                           0
Gender                        0
Platform                      0
Daily_Usage_Time (minutes)    0
Posts_Per_Day                 0
Likes_Received_Per_Day        0
Comments_Received_Per_Day     0
Messages_Sent_Per_Day         0
Dominant_Emotion              0
dtype: int64

----------------------------------------------------------------------
Validation Data:


User_ID                       0
Age                           0
Gender                        0
Platform                      0
Daily_Usage_Time (minutes)    0
Posts_Per_Day                 0
Likes_Received_Per_Day        0
Comments_Received_Per_Day     0
Messages_Sent_Per_Day         0
Dominant_Emotion              1
dtype: int64

* I can do 2 things to deal with null values:
  * Simple drop them as it is only 1 null value and might not affect our analysis.
  * Fill them with the mean, median, or mode depending on the value of the column.

# Exploratory Data Analysis (EDA)

* Let's start EDA with the training data.

In [10]:
train_df.head()

User_ID Age      Gender   Platform  Daily_Usage_Time (minutes)  \
0       1  25      Female  Instagram                       120.0   
1       2  30        Male    Twitter                        90.0   
2       3  22  Non-binary   Facebook                        60.0   
3       4  28      Female  Instagram                       200.0   
4       5  33        Male   LinkedIn                        45.0   

   Posts_Per_Day  Likes_Received_Per_Day  Comments_Received_Per_Day  \
0            3.0                    45.0                       10.0   
1            5.0                    20.0                       25.0   
2            2.0                    15.0                        5.0   
3            8.0                   100.0                       30.0   
4            1.0                     5.0                        2.0   

   Messages_Sent_Per_Day Dominant_Emotion  
0                   12.0        Happiness  
1                   30.0            Anger  
2                   20.0          Neutral  
3                   50.0          Anxiety  
4                   10.0          Boredom

* Coloum in our data:

In [11]:
# list of coloums in the training data
train_df.columns 

Index(['User_ID', 'Age', 'Gender', 'Platform', 'Daily_Usage_Time (minutes)',
       'Posts_Per_Day', 'Likes_Received_Per_Day', 'Comments_Received_Per_Day',
       'Messages_Sent_Per_Day', 'Dominant_Emotion'],
      dtype='object')

### Describe the training data

In [12]:
train_df.describe()

Daily_Usage_Time (minutes)  Posts_Per_Day  Likes_Received_Per_Day  \
count                 1000.000000    1000.000000             1000.000000   
mean                    95.950000       3.321000               39.898000   
std                     38.850442       1.914582               26.393867   
min                     40.000000       1.000000                5.000000   
25%                     65.000000       2.000000               20.000000   
50%                     85.000000       3.000000               33.000000   
75%                    120.000000       4.000000               55.000000   
max                    200.000000       8.000000              110.000000   

       Comments_Received_Per_Day  Messages_Sent_Per_Day  
count                1000.000000            1000.000000  
mean                   15.611000              22.560000  
std                     8.819493               8.516274  
min                     2.000000               8.000000  
25%                     8.000000              17.750000  
50%                    14.000000              22.000000  
75%                    22.000000              28.000000  
max                    40.000000              50.000000

### Age Distribution

In [13]:
train_df['Age'].isnull().sum()

0

In [14]:
train_df['Age'].unique()

array(['25', '30', '22', '28', '33', '21', '27', '24', '29', '31', '23',
       '26', '34', '35', '32', 'Male', 'Female', 'Non-binary',
       ' işte mevcut veri kümesini 1000 satıra tamamlıyorum:'],
      dtype=object)

* In the age coloum we have 4 irrigular values.
  * Male, Female, Non-binary, and other.
* ___Let's drop/fill these values first___

In [15]:
# removing the Male, Female, Non-binary, and işte mevcut veri kümesini 1000 satıra tamamlıyorum:

# Replace non-numeric values with NaN
train_df['Age'] = pd.to_numeric(train_df['Age'], errors='coerce')

# Handle NaN values
train_df['Age'].fillna(train_df['Age'].median(), inplace=True)


In [16]:
train_df['Age'].unique()

array([25., 30., 22., 28., 33., 21., 27., 24., 29., 31., 23., 26., 34.,
       35., 32.])

In [17]:
plt = px.histogram(train_df, x='Age', title='Age Distribution')
plt.show()

### Gender Distribution

In [18]:
train_df['Gender'].unique()

array(['Female', 'Male', 'Non-binary', '27', '24', '29', '33', '31', '22',
       '25', '28', '30', '23', '34', '26', '35', '21', '32', nan],
      dtype=object)

* We've the same issue here too. Mix of Gender with some number like in Age Coloum. Let's handle it.

In [19]:
# Function to replace numeric values with NaN
def clean_gender_column(gender_value):
    try:
        # Try converting the value to float, if it succeeds it's a numeric value
        float(gender_value)
        return np.nan
    except ValueError:
        # If conversion fails, it's a valid gender entry
        return gender_value

# Apply the function to the Gender column
train_df['Gender'] = train_df['Gender'].apply(clean_gender_column)

# Fill NaN values with a placeholder or drop them
# Here we fill NaN values with 'Unknown'
train_df['Gender'].fillna('Unknown', inplace=True)

# Verify the unique values after cleaning
print(train_df['Gender'].unique())


['Female' 'Male' 'Non-binary' 'Unknown']


In [20]:
train_df['Gender'].unique()

array(['Female', 'Male', 'Non-binary', 'Unknown'], dtype=object)

In [21]:
train_df['Gender'].value_counts()

Gender
Female        344
Male          332
Non-binary    248
Unknown        77
Name: count, dtype: int64

In [22]:
plt = px.histogram(train_df, x='Gender', title='Gender Distribution')
plt.show()

### Platform Distribution

In [23]:
train_df['Platform'].unique()

array(['Instagram', 'Twitter', 'Facebook', 'LinkedIn', 'Whatsapp',
       'Telegram', 'Snapchat', nan], dtype=object)

In [24]:
train_df['Platform'].value_counts()

Platform
Instagram    250
Twitter      200
Facebook     190
LinkedIn     120
Whatsapp      80
Telegram      80
Snapchat      80
Name: count, dtype: int64

* Let's Impute the nan with the mode

In [25]:
# filling with mode
train_df['Platform'].fillna(train_df['Platform'].mode()[0], inplace=True)

In [26]:
train_df['Platform'].unique()

array(['Instagram', 'Twitter', 'Facebook', 'LinkedIn', 'Whatsapp',
       'Telegram', 'Snapchat'], dtype=object)

In [27]:
train_df['Platform'].value_counts()

Platform
Instagram    251
Twitter      200
Facebook     190
LinkedIn     120
Whatsapp      80
Telegram      80
Snapchat      80
Name: count, dtype: int64

In [28]:
plt = px.histogram(train_df, x='Platform', title='Platform Distribution')
plt.show()

### Daily Usage Time (minutes) Distribution

In [29]:
plt = px.histogram(train_df, x='Daily_Usage_Time (minutes)', title='Daily Usage Time Distribution')
plt.show()

### Post Per Day Distribution

In [30]:
train_df['Posts_Per_Day'].unique()

array([ 3.,  5.,  2.,  8.,  1.,  4.,  6.,  7., nan])

In [31]:
# fill with mode
train_df['Posts_Per_Day'].fillna(train_df['Posts_Per_Day'].mode()[0], inplace=True)

In [32]:
plt = px.histogram(train_df, x='Posts_Per_Day', title='Posts Per Day Distribution')
plt.show()

### Likes Per Day Distribution

In [33]:
train_df['Likes_Received_Per_Day'].unique()

array([ 45.,  20.,  15., 100.,   5.,  60.,  30.,  25.,  10.,  80.,  35.,
        12.,  90.,  40.,  55.,  33.,   8.,  70.,  28.,  11.,  95.,  18.,
         9.,  85.,  38.,   6.,  13.,  75.,  27.,  88.,  22.,  78.,  29.,
        50.,  36.,  72.,  65., 110.,  14.,  17., 105.,  43.,  37.,  42.,
        48.,  21.,  24.,  23.,  83.,  nan])

In [34]:
# filling wih mode
train_df['Likes_Received_Per_Day'].fillna(train_df['Likes_Received_Per_Day'].mode()[0], inplace=True)

In [35]:
plt = px.histogram(train_df, x='Likes_Received_Per_Day', title='Posts Per Day Distribution')
plt.show()

### Comments Per Day Distribution

In [36]:
train_df['Comments_Received_Per_Day'].unique()

array([10., 25.,  5., 30.,  2., 15., 12.,  3., 20.,  7.,  4., 23., 18.,
       22., 14., 26.,  8., 19., 17., 11.,  6.,  9., 13., 40., 16., 35.,
       38., 28., 36., 33., nan])

In [37]:
# filling with mode
train_df['Comments_Received_Per_Day'].fillna(train_df['Comments_Received_Per_Day'].mode()[0], inplace=True)

In [38]:
train_df['Comments_Received_Per_Day'].unique()

array([10., 25.,  5., 30.,  2., 15., 12.,  3., 20.,  7.,  4., 23., 18.,
       22., 14., 26.,  8., 19., 17., 11.,  6.,  9., 13., 40., 16., 35.,
       38., 28., 36., 33.])

In [39]:
plt = px.histogram(train_df, x='Comments_Received_Per_Day', title='Posts Per Day Distribution')
plt.show()

### Messages Per Day Distribution

In [40]:
plt = px.histogram(train_df, x='Messages_Sent_Per_Day', title='Posts Per Day Distribution')
plt.show()

### Emotion Distribution

In [41]:
train_df['Dominant_Emotion'].unique()

array(['Happiness', 'Anger', 'Neutral', 'Anxiety', 'Boredom', 'Sadness',
       nan], dtype=object)

In [42]:
# fill with mode
train_df['Dominant_Emotion'].fillna(train_df['Dominant_Emotion'].mode()[0], inplace=True)

In [43]:
plt = px.pie(train_df, names='Dominant_Emotion', title='Dominant Emotion Distribution')
# adding the values to the pie section
plt.update_traces(textposition='inside', textinfo='percent+label')
plt.show()

## Relationship Between Variables

### Gender and Platform

In [44]:
# Group the data by gender and platform
grouped = train_df.groupby(['Gender', 'Platform'])

# Count the number of rows in each group
counts = grouped.size()

# Print the counts
print(counts)

Gender      Platform 
Female      Facebook      10
            Instagram    150
            LinkedIn      40
            Snapchat      28
            Twitter       66
            Whatsapp      50
Male        Facebook      38
            Instagram     76
            LinkedIn      48
            Telegram      50
            Twitter      102
            Whatsapp      18
Non-binary  Facebook     130
            Instagram     10
            LinkedIn      30
            Snapchat      40
            Telegram      18
            Twitter       20
Unknown     Facebook      12
            Instagram     15
            LinkedIn       2
            Snapchat      12
            Telegram      12
            Twitter       12
            Whatsapp      12
dtype: int64


In [45]:
plt = px.histogram(train_df, x='Gender', color='Platform', title='Platform by Gender Usage')
plt.show()

### Age and Gender

In [46]:
# grouping age with gender
grouped = train_df.groupby(['Age', 'Gender'])

# count the number of rows in each group
counts = grouped.size()

# print the counts  
print(counts)

Age   Gender    
21.0  Female        20
      Male          10
      Non-binary    26
22.0  Female        48
      Male           8
      Non-binary    18
23.0  Female        10
      Male           8
      Non-binary    30
24.0  Female        26
      Male          10
      Non-binary    28
25.0  Female        40
      Male          24
26.0  Female         8
      Male          28
      Non-binary    30
27.0  Female        28
      Male          44
      Non-binary    20
      Unknown       77
28.0  Female        36
      Male          28
      Non-binary    28
29.0  Female        28
      Male          30
      Non-binary    32
30.0  Female         8
      Male          40
31.0  Female         8
      Male          54
32.0  Female        28
      Male          10
33.0  Female        18
      Male          10
      Non-binary    28
34.0  Female        18
      Male          10
      Non-binary     8
35.0  Female        20
      Male          18
dtype: int64


In [47]:
plt = px.histogram(train_df, x='Age', color='Gender', title='Age by Gender')
plt.show()

### Gender and Platform VS Daily Usage Time (minutes)

In [48]:
plt = px.histogram(train_df, x='Posts_Per_Day', y='Platform' ,color='Gender', title='Posts Per Day by Gender')
plt.show()

### Gender VS Emotions

In [49]:
# checking the Gender againest the Dominant_Emotion
grouped = train_df.groupby(['Gender', 'Dominant_Emotion'])

# count the number of rows in each group
counts = grouped.size()

# print the counts  
print(counts)

Gender      Dominant_Emotion
Female      Anger                54
            Anxiety              54
            Boredom              30
            Happiness           102
            Neutral              56
            Sadness              48
Male        Anger                58
            Anxiety              56
            Boredom              54
            Happiness            66
            Neutral              46
            Sadness              52
Non-binary  Anger                10
            Anxiety              46
            Boredom              46
            Happiness            18
            Neutral              82
            Sadness              46
Unknown     Anger                 8
            Anxiety              14
            Boredom              10
            Happiness            15
            Neutral              16
            Sadness              14
dtype: int64


In [50]:
# ploting
plt = px.histogram(train_df, x='Gender', color='Dominant_Emotion', title='Dominant Emotion by Gender')
plt.show()

### Platform VS Emotions

In [51]:
# checking the Platformagainest the Dominant_Emotion
grouped = train_df.groupby(['Platform', 'Dominant_Emotion'])

# count the number of rows in each group
counts = grouped.size()

# print the counts
print(counts)

Platform   Dominant_Emotion
Facebook   Anxiety              50
           Boredom              40
           Neutral              70
           Sadness              30
Instagram  Anger                10
           Anxiety              30
           Happiness           171
           Neutral              20
           Sadness              20
LinkedIn   Anxiety              20
           Boredom              70
           Neutral              20
           Sadness              10
Snapchat   Anxiety              20
           Happiness            10
           Neutral              20
           Sadness              30
Telegram   Anger                10
           Anxiety              10
           Boredom              10
           Neutral              30
           Sadness              20
Twitter    Anger                80
           Anxiety              20
           Boredom              20
           Happiness            10
           Neutral              20
           Sadness         

In [52]:
Plt = px.histogram(train_df, x='Platform', color='Dominant_Emotion', title='Dominant Emotion by Platform')
Plt.show()

In [53]:
# Create a contingency table
contingency_table = pd.crosstab(train_df['Platform'], train_df['Dominant_Emotion'])

# Plot the heatmap
fig = px.imshow(contingency_table, title='Platform vs Dominant Emotion Heatmap')
fig.show()

### Time Spent VS Emotions

In [54]:
# Daily Usage time by Dominant_Emotion
plt = px.histogram(train_df, x='Daily_Usage_Time (minutes)', color='Dominant_Emotion', title='Time Usage by Dominant Emotion')
plt.show()

### Likes Received VS Emotions
* This is the last realation I'll be looking in this data.

In [55]:
plt = px.histogram(train_df, x='Likes_Received_Per_Day', color='Dominant_Emotion', title='Likes Received vs Dominant Emotion')
plt.show()

# Model Training

In [56]:
# Import the necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report

In [57]:
# Checking for missing values
train_df.isnull().sum()

User_ID                       0
Age                           0
Gender                        0
Platform                      0
Daily_Usage_Time (minutes)    1
Posts_Per_Day                 0
Likes_Received_Per_Day        0
Comments_Received_Per_Day     0
Messages_Sent_Per_Day         1
Dominant_Emotion              0
dtype: int64

In [58]:
# Let's drop the missing values
train_df.dropna(inplace=True)

In [59]:
# Define features and target
X = train_df.drop(columns=['Dominant_Emotion', 'User_ID']) # Features
y = train_df['Dominant_Emotion'] # Target

# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [60]:
# Define preprocessing steps. Seeperating the numerical and categorical features
numeric_features = ['Daily_Usage_Time (minutes)', 'Posts_Per_Day', 'Likes_Received_Per_Day', 'Comments_Received_Per_Day', 'Messages_Sent_Per_Day']
categorical_features = ['Age', 'Gender', 'Platform']

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)])

### Random Forest Classifier

`Random Forest is an ensemble learning method that constructs multiple decision trees during training and outputs the mode of the classes (classification). It improves predictive performance and reduces overfitting compared to a single decision tree by averaging the results of many trees.`

In [61]:
# Random Forest Classifier
rf_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                              ('classifier', RandomForestClassifier(random_state=42))])

rf_pipeline.fit(X_train, y_train)
y_pred_rf = rf_pipeline.predict(X_test)

print("Random Forest Classifier Report:")
print(classification_report(y_test, y_pred_rf))

Random Forest Classifier Report:
              precision    recall  f1-score   support

       Anger       1.00      1.00      1.00        29
     Anxiety       1.00      0.97      0.99        39
     Boredom       1.00      1.00      1.00        31
   Happiness       1.00      1.00      1.00        35
     Neutral       0.98      1.00      0.99        42
     Sadness       1.00      1.00      1.00        24

    accuracy                           0.99       200
   macro avg       1.00      1.00      1.00       200
weighted avg       1.00      0.99      0.99       200



### XGBoost Classifier

`XGBoost (Extreme Gradient Boosting) is a scalable and efficient implementation of gradient boosting algorithms. It builds an ensemble of trees in a sequential manner, where each tree tries to correct the errors of the previous trees. XGBoost is known for its high performance and speed, making it popular for structured or tabular data tasks in machine learning competitions and real-world applications.`


In [62]:
from sklearn.preprocessing import LabelEncoder
# Encode target labels as integers
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

In [63]:
# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

# XGBoost Classifier
xgb_pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', XGBClassifier(random_state=42))])

xgb_pipeline.fit(X_train, y_train)
y_pred_xgb = xgb_pipeline.predict(X_test)

In [64]:
# Decode the predicted labels back to original string labels for the report
y_test_decoded = label_encoder.inverse_transform(y_test)
y_pred_xgb_decoded = label_encoder.inverse_transform(y_pred_xgb)

print("XGBoost Classifier Report:")
print(classification_report(y_test_decoded, y_pred_xgb_decoded))

XGBoost Classifier Report:
              precision    recall  f1-score   support

       Anger       1.00      1.00      1.00        29
     Anxiety       1.00      0.92      0.96        39
     Boredom       0.94      1.00      0.97        31
   Happiness       1.00      1.00      1.00        35
     Neutral       0.98      1.00      0.99        42
     Sadness       1.00      1.00      1.00        24

    accuracy                           0.98       200
   macro avg       0.99      0.99      0.99       200
weighted avg       0.99      0.98      0.98       200



## Neural Network

In [65]:
from sklearn.compose import ColumnTransformer
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.utils import to_categorical

2024-05-23 11:23:00.454061: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-05-23 11:23:00.454295: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-05-23 11:23:00.579179: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [66]:
# Define features and target
X = train_df.drop(columns=['Dominant_Emotion', 'User_ID'])
y = train_df['Dominant_Emotion']

# Encode target labels as integers
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

# One-hot encode the target labels for neural network
y_onehot = to_categorical(y_encoded)

# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y_onehot, test_size=0.2, random_state=42)

# Define preprocessing steps for features
numeric_features = ['Daily_Usage_Time (minutes)', 'Posts_Per_Day', 'Likes_Received_Per_Day', 'Comments_Received_Per_Day', 'Messages_Sent_Per_Day']
categorical_features = ['Age', 'Gender', 'Platform']

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)])

# Preprocess the features
X_train = preprocessor.fit_transform(X_train)
X_test = preprocessor.transform(X_test)

#### Build the neural network

In [67]:
# Build the neural network
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(y_onehot.shape[1], activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.1)

Epoch 1/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 2s 19ms/step - accuracy: 0.2200 - loss: 1.7807 - val_accuracy: 0.4750 - val_loss: 1.4713
Epoch 2/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3372 - loss: 1.5990 - val_accuracy: 0.6250 - val_loss: 1.3293
Epoch 3/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.4466 - loss: 1.4568 - val_accuracy: 0.6125 - val_loss: 1.2476
Epoch 4/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3941 - loss: 1.4792 - val_accuracy: 0.6000 - val_loss: 1.1823
Epoch 5/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5012 - loss: 1.3576 - val_accuracy: 0.6500 - val_loss: 1.1091
Epoch 6/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5192 - loss: 1.2723 - val_accuracy: 0.7125 - val_loss: 1.0398
Epoch 7/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5414 - loss: 1.2374 - val_accuracy: 0.7500 - val_loss: 0.9670
Epoch 8/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5476 - loss: 1.2134 - val_accuracy: 0.7625 - val_loss

In [68]:
# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Accuracy: {accuracy:.2f}')

# Make predictions
y_pred = model.predict(X_test)
y_pred_classes = np.argmax(y_pred, axis=1)
y_test_classes = np.argmax(y_test, axis=1)

# Decode the predicted and true labels back to original string labels for the report
y_pred_labels = label_encoder.inverse_transform(y_pred_classes)
y_test_labels = label_encoder.inverse_transform(y_test_classes)

7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9421 - loss: 0.1988 
Accuracy: 0.95
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step


In [69]:

# Print classification report
print("Neural Network Classifier Report:")
print(classification_report(y_test_labels, y_pred_labels))

Neural Network Classifier Report:
              precision    recall  f1-score   support

       Anger       1.00      0.93      0.96        29
     Anxiety       0.97      0.95      0.96        39
     Boredom       0.97      0.97      0.97        31
   Happiness       0.97      1.00      0.99        35
     Neutral       0.89      0.95      0.92        42
     Sadness       0.96      0.92      0.94        24

    accuracy                           0.95       200
   macro avg       0.96      0.95      0.96       200
weighted avg       0.96      0.95      0.96       200



# Insights:
1. `Age Distribution`: We can see that the majority of the users are in all ages between 21 and 35. The higest count is in the age group of 27.
2. `Gender Distribution`: We've 4 Values in the Gender Column. _Male_, _Female_, _Non-Binary_, _Unknown_.
   1. Male: 332,
   2. Female: 344,
   3. Non-Binary: 248,
   4. Unknown: 77
3. `Platform Distribution`: We can see that the majority of the users use Instagram followed by Twitter and Facebook.
   1. Instagram    251
   2. Twitter      200
   3. Facebook     190
   4. LinkedIn     120
   5. Whatsapp      80
   6. Telegram      80
   7. Snapchat      80
4. `Daily_Usage_Time (minutes)`: The Daily Usage Time is from 40 to 200 minutes. Most count is in range of 60 to 90 minutes. While the Average time is 95 min.M
5. `Post per Day:` The Minimum number of post is 1 and Maximum is 8, while the average is 3.32 posts. (Most Count is 2)
6. `Likes Received per Day:` The Minimum number of like received is 5 and Highest is 110, while the average is 39.89.
7. `Comments Received per Day:` The Minimum number of comment received is 2 and maximum is 40 while the average if 15.61.
8. `Messages Sent per Day:` The Minimum number of messages sent is 8 and maximum is 50 while the average if 22.56.
9. `Emotions`: We've total 6 emotions: 
   1.  Happiness: 20.1%
   2.  Anger: 13%
   3.  Neutral: 20%
   4.  Anxiety: 17
   5.  Boredom: 14%
   6.  Sadness: 16%
10. `Gender VS Platform`: 
    1.  _Instagram_ is most usage app amoung `female` while facebook is the least one.
    2.  _Twitter_ is the most usage app amoung `male` followed by instagram. While Whatsapp is the least one.
    3.  _Facebook_ is the most usage app amoung `non-binary`.
11. ` Post Per Day By Gender`
    1.   Instagram:
         1.   Total Posts Per Day(Female): 885
         2.   Total Posts Per Day(Female): 443
    2. Twitter/X:
       1.  Total Posts Per Day(Female): 226
       2.  Total Posts Per Day(Female): 351
    3. LinkedIn:
       1. Total Posts Per Day(Female): 60
       2. Total Posts Per Day(Female): 58
12. `Dominent Emotion by Gender`: 
    1.  __Female:__ The dominent emotion is _Happniess_ and the count is 102. While _Anger_, _Nuetral_, _Anxiety_ and _Sadness_ are the other emotions with the almost same count range 56 - 48. The count of _Boredom_ is 30 which is the least count amoung other.
    2.  __Male__: The dominent emotion is _Happniess_ and the count is 66. While _Anger_, _Nuetral_, _Anxiety_, _Boredom_and _Sadness_ are the other emotions with the almost same count range 58 - 46.
    3.  __Non-Binary__: The dominent emotion is _Neutral_ and the count is 82. While _Anxiety_, _Sadness_ and _Boredom_ are the other emotions with the almost same count 46. Anger is the leaset count (10) among the other.
13. `Dominant_Emotion` is strongly correlated with `Platform`. 
    1.  _Happiness_ is the dominant emotion in the _Instagram_ platform. While _Anger_ is the least count.
    2.  _Anger_ is the dominant emotion in the _Twitter_ platform and _Whatsapp_. While _Hapniess_ is the least count in both.
    3.  _Sadness_ is the also dominant emotion in the _Twitter_ platform and _Snapchat_.
    4.  _Neutral_ is the dominant emotion in the _Facebook_ platform and _Telegram_ platform.
    5.  _Boredom_ is the dominant emotion in the _LinkedIn_ platform and _Facebook_
    6.  _Anixity_ is the also dominant emotion in the _Facebook_ and other platforms.
14. `Daily_Usage_Time (minutes)` and `Dominant_Emotion` are strongly correlated.
    1.  `200 minutes` is related with `Anxiety` emotion.
    2.  `From 140 to 190 minutes` is related with `Happiness` emotion.
    3.  `Anger` is commonly seens in the range of `60 to 120 minutes`.
    4.  Other Emotions are also in the range of `40 to 130 minutes`
15. `Likes Received Per Day` is strongly correlated with `Dominant_Emotion`.
    1.  Upon looking at the plot, we can see that the `Happiness` emotion is triggered by the `Likes_Received_Per_Day` Ranging from 60 to 109 Likes.
    2.  While the `Anxiety` emotion is also triggered by the `Likes_Received` Ranging from 110 to 114, which could means more the likes received is causes the Anxiety.

    


## About Model:
`Random Forest Classifier`: __accuracy: 99%__
The Random Forest Classifier is performing exceptionally well on this dataset, with nearly perfect precision, recall, and f1-scores for most classes and an overall accuracy of 99%. This indicates that the model is highly effective at predicting the dominant emotion based on the given features.

`XGBoost Classifier`:__accuracy: 98%__
The XGBoost Classifier is performing very well on this dataset, with high precision, recall, and f1-scores for most classes and an overall accuracy of 98%. This indicates that the model is highly effective at predicting the dominant emotion based on the given features, though it shows slightly less precision in predicting 'Anxiety' compared to other emotions.

`Neural Network`: __accuracy: 94%__
The neural network classifier is highly effective in identifying various emotions, achieving a high overall accuracy and strong class-wise performance. Focusing on improving recall for Boredom and Sadness and balancing precision and recall for Neutral could further enhance the classifier's robustness.

<h1><b><p style="background-image: url(https://i.postimg.cc/0Qwf8YX6/2927262.jpg);font-family:camtasia;font-size:110%;color:white;text-align:center;border-radius:15px 50px; padding:7px; border:solid 2px #09375b; box-shadow: 10px 10px 10px #042b4c">THE END <br>
FOLLOW | UPVOTE | COMMENT </p></b></h1>